docs/robotframework-tutorial.rst

```diff
@@ -0,0 +1,382 @@
+second argument is the object name (eg: ``MyObject__c``).
+
+Our test is already using ``Go to page``, so our keyword should
+already be available to us once we've gone to that page.
+
+
+Part 5: Adding test data
+========================
+
+We want to be able to test that when we click on one of our custom
+objects on the listing page that it will take us to the detail page
+for that object. To do that, our test needs some test data. While that
+can be very complicated in a real-world scenario, for simple tests we
+can use the salesforce API to create test data when the suite first
+starts up.
+
+To create the data when the suite starts, we can add a ``Suite Setup``
+in the settings section of the test. This takes as an argument the
+name of a keyword. In our case we're going to create a custom keyword
+right in the test to add some test data for us.
+
+It is not necessary to do it in a setup. It could be a step in an
+individual test case, for example. However, putting it in the ``Suite
+Setup`` guarantees it will run before any tests in the same file are
+run.
+
+Open up ``MyObject.robot`` and add the following just before ``***
+Test Cases ***`:
+
+.. code-block:: console
+
+    *** Keywords ***
+    Create test data
+        [Documentation]
+        ...  Creates a MyObject record named "Leeroy Jenkins"
+        ...  if one doesn't exist
+
+        # Check to see if the record is already in the database,
+        # and return if it already exists
+        ${status}  ${result}=  Run keyword and ignore error  Salesforce get  MyObject__c  Name=Leeroy Jenkins
+        Return from keyword if  '${status}'=='PASS'
+
+        # The record didn't exist, so create it
+        Log  creating MyObject object with name 'Leeroy Jenkins'  DEBUG
+        Salesforce Insert  MyObject__c  Name=Leeroy Jenkins
+
+We also need to modify our ``Suite Setup`` to call this keyword in
+addition to calling the ``Open Test Browser`` keyword. Since ``Suite
+Setup`` only accepts a single keyword, we can use the built-in keyword
+``Run keywords`` to run more than one keyword in the setup.
+
+Change the suite setup to look like the following, again using robot's
+continuation characters to spread the code across multipe rows for
+readability.
+
+.. note::
+
+    Note: it is critical that you use all caps for ``AND``, as
+    that's the way robot knows where one keyword ends and the next
+    begins.
+
+.. code-block:: console
+
+    Suite Setup     Run keywords
+    ...  Create test data
+    ...  AND  Open test browser
+
+Notice that our ``Suite Teardown`` calls ``Delete records and
+close browser``. The _records_ in that keyword refers to any data
+records created by ``Salesforce Insert``. This makes it possible to
+both create and later clean up temporary data used for a test.
+
+It is important to note that the suite teardown isn't guaranteed to run
+if you forcibly kill a running robot test. That is why we added the
+check in the ``Create test data`` keyword.
```

Comment:
I don't quite follow this sentence. We added the check so that if the test is run twice against the same org it will not fail due to something like a duplicate rule?

Reply:
The scenario is this: the test adds "Leeroy Jenkins". If you forcibly stop the test before the teardown runs, that record will remain in the DB. The next time you run the test, without the check you'll end up with two records with that name. While essentially harmless in a scratch org, and harmless in this particular scenario, we don't want to be littering the database with unnecessary objects.

Reply:
I've reworded this paragraph slightly. Thanks for the comment!